### PR TITLE
feat: gitlab optimization with caching fix

### PR DESCRIPTION
**Pull request by rarkins**

### CODE REVIEW

File: lib/platform/gitlab/index.ts [hunk 1]
```diff
@@ -613,15 +613,19 @@ const mapPullRequests = (pr: {
   createdAt: pr.created_at,
 });
 
-export async function getPrList() {
```
Comment (ViceIce) [inline lib/platform/gitlab/index.ts]: I don't think that this will fix the problem.

**Comment by rarkins** [comment]
I think it’s still the correct way we should be caching the list. But if there are problems with findPr/getBranchPr then we should address them also.

**Comment by ViceIce** [comment]
OK, i will debug that a little today evening or tomorrow morning. I'm pretty sure that it has something todo with the `query` object of type `URLSearchParams`.

**Comment by ViceIce** [comment]
https://github.com/sindresorhus/got#searchparams

I think `query` should be `searchParams` ?

**Comment by ViceIce** [comment]
No, is only renamed in got v10 but working. in v11 only `searchParams` will work.

**Comment by leipert** [comment]
Sorry, in dry run this worked perfectly for me 😢 

Is it possible to set `useCache: false` in the two requests?

**Comment by leipert** [comment]
> I'm pretty sure that it has something todo with the query object of type URLSearchParams.

This could actually be, could it be that: https://github.com/renovatebot/renovate/blob/master/lib/util/got/cache-get.js#L24 is not working correctly with the `URLSearchParams`? And then we basically land in an infinite loop where we always request the next page?

**Comment by leipert** [comment]
Yeah, I think looking at this and https://github.com/renovatebot/renovate/pull/3936#issuecomment-509504775, I think the following happens:

1. Request to `const res = await api.get(`projects/${config.repository}/merge_requests`, { query, paginate: true });`.
2. Response with `X-Next-Page: 2` link header.
3. Response gets cached.
4. Next pagination request goes out to the same url, get's served the cached result, link header is still the same
5. rinse and repeat

**Comment by rarkins** [comment]
We definitely should include the query Params in our cache key either way though, as it seems they’re not stringified into the href by the time the request bits our cache logic.

**Comment by leipert** [comment]
Ah looking at it again, everything is fine with caching, unfortunately in the following line we are destroying pagination by overwriting it with the queryParams we are setting 🤦‍♂ 

https://github.com/renovatebot/renovate/blob/cbf695bb392771eadf1e80e19f05e50853fc6c16/lib/platform/gitlab/gl-got-wrapper.ts#L24

**Comment by leipert** [comment]
@rarkins Fixing it with this: https://github.com/renovatebot/renovate/pull/4038

**Comment by leipert** [comment]
The diff is essentially:
```
diff --git a/lib/platform/common.ts b/lib/platform/common.ts
index 512ac0e8..a8703c70 100644
--- a/lib/platform/common.ts
+++ b/lib/platform/common.ts
@@ -4,7 +4,6 @@ export interface IGotApiOptions {
   useCache?: boolean;
   hostType?: string;
   body?: any;
-  query?: string | { [key: string]: string | number } | URLSearchParams;
 }
 
 export interface IGotApi<TOptions extends object = any> {
diff --git a/lib/platform/gitlab/index.ts b/lib/platform/gitlab/index.ts
index 87d59e7e..0e55fb6d 100644
--- a/lib/platform/gitlab/index.ts
+++ b/lib/platform/gitlab/index.ts
@@ -211,9 +211,9 @@ export async function getBranchPr(branchName: string) {
     per_page: '100',
     state: 'opened',
     source_branch: branchName,
-  });
-  const urlString = `projects/${config.repository}/merge_requests`;
-  const res = await api.get(urlString, { query, paginate: true });
+  }).toString();
+  const urlString = `projects/${config.repository}/merge_requests?${query}`;
+  const res = await api.get(urlString, { paginate: true });
   logger.debug(`Got res with ${res.body.length} results`);
   let pr: any = null;
   res.body.forEach((result: { source_branch: string }) => {
@@ -617,15 +617,15 @@ async function fetchPrList() {
   const query = new URLSearchParams({
     per_page: '100',
     author_id: `${authorId}`,
-  });
-  const urlString = `projects/${config.repository}/merge_requests`;
-  const res = await api.get(urlString, { query, paginate: true });
+  }).toString();
+  const urlString = `projects/${config.repository}/merge_requests?${query}`;
+  const res = await api.get(urlString, { paginate: true });
   return res.body.map(mapPullRequests);
 }
 
-export function getPrList() {
+export async function getPrList() {
   if (!config.prList) {
-    config.prList = fetchPrList();
+    config.prList = await fetchPrList();
   }
   return config.prList;
 }
```

**Comment by ViceIce** [comment]
@leipert Have you done some real tests with #4038? A `dryRun` should be enough.

**Comment by rarkins** [comment]
I'm guessing that to properly test pagination we need someone to run it where more than one page is returned. Or temporarily set the pagination to 1 just for testing.

**Comment by ViceIce** [comment]
@rarkins I can do a real test when I'm back at home at ~16:30

**Comment by leipert** [comment]
@rarkins / @ViceIce Yes, I tested it with https://gitlab.com/leipert-projects/renovate-gitlab-bot locally in not-dry-run mode.

**Comment by leipert** [comment]
It runs against https://gitlab.com/gitlab-org/gitlab-ce which has two pages of MRs with my personal token.

**Comment by leipert** [comment]
e.g. https://gitlab.com/api/v4/projects/gitlab-org%2Fgitlab-ce/merge_requests?author_id=181229&per_page=100

**Pull request merged**